### DISCUSSION TITLE
Download private key from crypto token

### DISCUSSION THREAD

**Discussion by sharbich** [Keyfactor/ejbca-ce#721] [Q&A]
Hello, I cannot download the private key of my CA using the following command:
`ejbca.sh ca exportca HarbichCA HarbichCA.p12`
I selected the crypto token in the menu for soft crypto tokens in the admin interface -> crypto tokens, switched to edit mode, checked the box "Allow export of private keys" and saved. However, this is not being adopted. Why?
![Bildschirmfoto vom 2024-11-26 12-25-32](https://github.com/user-attachments/assets/3d7c43f2-4d8d-405e-87eb-9306a7e1f063)
Greetings from Stefan Harbich

**Comment by primetomas** [comment]
The command is names exportca because it is for CAs, not crypto tokens. You'll create a CA using a crypto token and then the exported P12 will be containing the keys that the CA is configured to use. 
Crypto Tokens as such can unfortunately only be fully exported with [ConfigDump](https://docs.keyfactor.com/ejbca/latest/configdump-tool).

**Comment by sharbich** [comment]
root@dsme01:/opt/ejbca/bin# ./ejbca.sh ca exportca HarbichCA HarbichCA.p12
Setting SignatureKeyAlias to SignatureKeyAlias
Setting EncryptionKeyAlias to EncryptionKeyAlias
Enter keystore password: 

Exception in thread "main" javax.ejb.EJBException: org.cesecore.keys.token.CryptoTokenAuthenticationFailedException: PKCS12 key store mac invalid - wrong password or corrupted file.
Unfortunately, I don't know the password anymore. Can I read the password?

**Comment by sharbich** [comment]
I only want to export the CA's public and private key?

**Reply by primetomas** [reply]
If you have created keys for a CA/Crypto Token but forgot the password (not the default one used, which would have been insecure from the start) it is nothing we can help with unfortunately. PKI is based on cryptography without backdoors, and the same with the encryption of PKCS#12 files in the database. PKCS#12 is a standard, and currently it utilizes 3DES and password based key generation. Neither of these are crackable. The only thing possible is to guess/brute force your password if it was a weak one.

Default values foo123 or !secret! was provided before, but the ability to use a "default" value was removed some time ago for security reasons (i.e no-one should use any default value).

**Reply by primetomas** [reply]
Aha, I see that you have auto-activation enabled right? In that case the password is stored in the database, but encrypted. It is a bit of work to try to decrypt that, but it is possible.

**Reply by sharbich** [reply]
> It is a bit of work to try to decrypt that, but it is possible.
Can you give me a tip how I can decrypt the password?

**Reply by sharbich** [reply]
I see in the ejbca.properties the folling entry "ca.cmskeystorepass=<password>. Is this the keystore password?

**Reply by sharbich** [reply]
I'm happy. I found the password in the "cesecore.properties" file. How can I change the password?

**Reply by primetomas** [reply]
You can change crypto token password in the UI or using the CLI.
bin/ejbca.sh cryptotoken setpin --help
